crates/neon-runtime/src/napi/task.rs

```diff
@@ -1,7 +1,329 @@
+    fn drop(&mut self) {
+        unsafe {
+            napi::napi_delete_async_work(self.env, self.work);
+        }
+    }
+}
+
+unsafe extern "C" fn perform_work<Data, Execute, Output, Complete>(
+    _: napi::napi_env,
+    task: *mut c_void,
+)
+where
+    Execute: FnOnce(Data) -> Output,
+{
+    let task = &mut *(task as *mut AsyncTask<Data, Execute, Output, Complete>);
+    let execute = task.execute.take().unwrap();
+    let data = task.data.take().unwrap();
+
+    task.output.replace(execute(data));
+}
+
+// Complete async work on the main thread
+unsafe extern "C" fn complete_work<Data, Execute, Output, Complete>(
+    env: napi::napi_env,
+    status: napi::napi_status,
+    task: *mut c_void,
+)
+where
+    Execute: FnOnce(Data) -> Output,
+    Complete: FnOnce(napi::napi_env, Output) -> Option<Local>,
+{
+    // Unbox the task first to ensure it is dropped
+    let mut task = NapiAsyncWork::<Data, Execute, Output, Complete>::from_raw(env, task);
+
+    // Neon does not support cancelling tasks but there may be other ways for
+    // tasks to become cancelled.
+    // Must come after `task` is unboxed to prevent memory leaks.
+    if status == napi::napi_status::napi_cancelled {
+        return;
+    }
+
+    // Run the provided `complete` function
+    let output = task.task.output.take().unwrap();
+    let complete = task.task.complete.take().unwrap();
+    let output = complete(env, output);
+
+    // Determine if an exception has been thrown
+    let mut result = MaybeUninit::uninit();
+    let status = napi::napi_is_exception_pending(env, result.as_mut_ptr());
+    let is_exception = {
+        assert_eq!(status, napi::napi_status::napi_ok);
+
+        result.assume_init()
+    };
+
+    // An exception has been thrown, call with `callback(err)`
+    let mut argv = if is_exception {
+        let mut result = MaybeUninit::uninit();
+        let status = napi::napi_get_and_clear_last_exception(
+            env,
+            result.as_mut_ptr(),
+        );
+
+        let exception = {
+            assert_eq!(status, napi::napi_status::napi_ok);
+
+            result.assume_init()
+        };
+
+        [exception, undefined(env)]
+
+    // Completed successfully, call with `callback(null, output)`
+    } else {
+        [null(env), output.unwrap_or_else(|| undefined(env))]
+    };
+
+    let mut result = MaybeUninit::uninit();
+    let status = napi::napi_get_reference_value(
+        env,
+        task.callback.as_raw(),
+        result.as_mut_ptr(),
+    );
+
+    let callback = {
+        assert_eq!(status, napi::napi_status::napi_ok);
+
+        result.assume_init()
+    };
+
+    assert_eq!(
+        napi::napi_call_function(
+            env,
+            global(env),
+            callback,
+            2,
+            argv.as_mut_ptr(),
+            MaybeUninit::uninit().as_mut_ptr(),
+        ),
+        napi::napi_status::napi_ok,
+    );
+}
+
+pub unsafe fn schedule<Data, Execute, Output, Complete>(
```

Comment:
The bulk of the implementation is contained in this file. There are several raw pointers that can leak if not handled properly:

* `napi::napi_ref`. Reference counted persistent handle to the JavaScript callback.
* `napi::napi_async_work`. Data represented the task to be submitted to the queue.
* `*mut AsyncWork<_, _, _, _>`. Rust struct for tracking state across suspension points.

Additionally, there are several points of failure that may occur when some or all of the above have been created. As an extra precaution, two RAII guards were created to ensure data is freed appropriately.

* `NapiCallback`. Simple wrapper for the `napi::napi_ref` JavaScript callback.
* `NapiAsyncWork`. Wrapper for all manually managed data above, inclusive of `NapiCallback`.

Only after successfully submitting the task to the work queue is `NapiAsyncWork` forgotten with `std::mem::forget`. Up until this point, early returns, including `panic` will execute `Drop` and free manually managed data.

Two assumptions are made:
* The `Execute` closure will not panic. If it does, memory will leak. This is undesired, but is still considered *safe*.
* Node will execute the task if and only if the call to ` napi::napi_queue_async_work` returns a status of `napi::napi_status::napi_ok`. If the task is executed, despite returning a failure status, a double free will occur. This *is* undefined behavior and will likely result in a segfault. This is acceptable because it would be considered a _bug_ in n-api.